### PR TITLE
Deepak saini119/fast metrics

### PR DESCRIPTION
**Pull request by DeepakSaini119**

### CODE REVIEW

File: xclib/evaluation/xc_metrics.py
Comment (anirudhb11) [inline xclib/evaluation/xc_metrics.py]: @DeepakSaini119 would it be possible to instead have a version of recall that computes 1..k, I think it would be pretty similar to how you have implemented precision. Given we a set of m documents we'll have a `mxk` array of `cum_intersections` we would just need to divide this by a `mx1` array which has the number of ground truth for each elt. Am I missing something?

Comment (DeepakSaini119) [reply xclib/evaluation/xc_metrics.py]: I think adding such a thing would be wasteful. Recall is generally required for large values for k. For instance, if R@100 is required, R@99 would be useless in most cases. But for uniformity, we can have a 1...k version of recall as well For this, we can go ahead as you suggest

**Comment by kunaldahiya** [comment]
@anirudhb11 I debug the code and 'serial' top-k is the main bottleneck. I have modified the code to find top-k using numba (default; parallel version) or cython (original; may be used if numba fails on some system). All evaluation functions can make use of it and should be faster now - no need to repeat implementation except may be for recall_at_just_k. 

Can you please take a look?

**Comment by DeepakSaini119** [comment]
We also need versions of functions which don't need the indices, scores to be first converted to csr_matrix as this step is extrmely slow for large datasets

**Comment by DeepakSaini119** [comment]
> @anirudhb11 I debug the code and 'serial' top-k is the main bottleneck. I have modified the code to find top-k using numba (default; parallel version) or cython (original; may be used if numba fails on some system). All evaluation functions can make use of it and should be faster now - no need to repeat implementation except may be for recall_at_just_k.
> 
> Can you please take a look?

We also need versions of functions which don't need the indices, scores to be first converted to csr_matrix as this step is extremely slow for large datasets

**Comment by kunaldahiya** [comment]
@DeepakSaini119 This is already implemented in the code - just pass the indices and scores as a dict with keys 'indices' and 'scores'.  You can also use 'sorted' flag if the scores are already sorted per row.

**Review by sonumehta — APPROVED** [review]
Please merge this PR once Kunal commits documentation related changes. 
Everything else looks good to me

File: xclib/evaluation/xc_metrics.py
Comment (kunaldahiya) [inline xclib/evaluation/xc_metrics.py]: Can we please ask the user for the pad value instead of hardcoding it?

File: xclib/evaluation/xc_metrics.py
Comment (kunaldahiya) [inline xclib/evaluation/xc_metrics.py]: Can we please use a better name now? How about 'recall_at_k'?

File: xclib/evaluation/xc_metrics.py
Comment (kunaldahiya) [inline xclib/evaluation/xc_metrics.py:840]: What is this trying to do? Compute recall with given top indices and their corresponding scores?

File: xclib/examples/fast_evaluate.py
Comment (kunaldahiya) [inline xclib/examples/fast_evaluate.py:30]: We need a better way to supply k. People compute recall at higher values (10/20/50/100 etc.,) as compared to precision (1-5). Btw, nDCG is missing from the example.

What do you think?

File: xclib/examples/fast_evaluate.py
Comment (kunaldahiya) [inline xclib/examples/fast_evaluate.py]: Can we please include the command in a way that it doesn't include our names?

File: xclib/evaluation/xc_metrics.py
Comment (kunaldahiya) [inline xclib/evaluation/xc_metrics.py]: Use 'getnnz(axis=1)' method?

File: xclib/evaluation/xc_metrics.py
Comment (kunaldahiya) [inline xclib/evaluation/xc_metrics.py]: User should be able to compute recall@gt and microR@gt separately. Moreover, what if user has predictions as 'csr_matrix' matrix or as 2 nd-arrays (indices and values)? Then old functions including precision, recall, and nDCG etc., would work but this won't. 

Support multiple formats (see https://github.com/kunaldahiya/pyxclib/blob/74da00fef66cddd12e40406f723df07eb4d9a120/xclib/evaluation/xc_metrics.py#L230)?

**Comment by kunaldahiya** [comment]
@anirudhb11 @sonumehta There are some possible modifications for consistency and user-friendliness - added comments. Please take a look.

File: xclib/evaluation/xc_metrics.py
Comment (anirudhb11) [reply xclib/evaluation/xc_metrics.py]: pad value index is now taken as an input from the user (as a command line argument): (refer [here](https://github.com/kunaldahiya/pyxclib/blob/fc3bb71c57283ea28ef8f9377572534102a10918/xclib/examples/fast_evaluate.py#L56)) and then passed as an argument to this block of code.

File: xclib/evaluation/xc_metrics.py
Comment (anirudhb11) [reply xclib/evaluation/xc_metrics.py]: Renamed to `recall_at_k`

File: xclib/examples/fast_evaluate.py
Comment (anirudhb11) [reply xclib/examples/fast_evaluate.py:30]: - k is now passed as a command line argument (refer [here](https://github.com/kunaldahiya/pyxclib/blob/fc3bb71c57283ea28ef8f9377572534102a10918/xclib/examples/fast_evaluate.py#L57)). 
- nDCG has now been added (refer [here](https://github.com/kunaldahiya/pyxclib/blob/DeepakSaini119/fast_metrics/xclib/examples/fast_evaluate.py#L39)
- Although metrics like P@100 will be very low for the datasets we use, we wanted to verify for a particular `k` is the behavior across metrics the same.

File: xclib/examples/fast_evaluate.py
Comment (anirudhb11) [reply xclib/examples/fast_evaluate.py]: Modified to [this](https://github.com/kunaldahiya/pyxclib/blob/DeepakSaini119/fast_metrics/xclib/examples/fast_evaluate.py#L50)

File: xclib/evaluation/xc_metrics.py
Comment (anirudhb11) [reply xclib/evaluation/xc_metrics.py]: Done, (refer [here](https://github.com/kunaldahiya/pyxclib/blob/DeepakSaini119/fast_metrics/xclib/evaluation/xc_metrics.py#L753))

File: xclib/evaluation/xc_metrics.py
Comment (anirudhb11) [reply xclib/evaluation/xc_metrics.py]: Added 2 new functions `micro_recall_at_gt` and `recall_at_gt` along with their corresponding helper functions with `_` prefix, this now supports for `csr_matrix` and `np.2darray`, like other functions.

**Comment by kunaldahiya** [comment]
@anirudhb11 Looks good. Shall I go ahead and merge?

**Comment by sonumehta** [comment]
> @anirudhb11 Looks good. Shall I go ahead and merge?
Yes Kunal, Please merge. We were waiting for your final review. Everything else looks good

**Pull request merged**
